### PR TITLE
KAFKA-6168 Connect Schema comparison is slow for large schemas

### PR DESCRIPTION
**Pull request by tedyu**
Re-arrange order of comparisons in equals() to evaluate non-composite fields first
Cache hash code

### CODE REVIEW

**Comment by asfgit** [comment]
FAILURE
 No test results found.
--none--

**Comment by asfgit** [comment]
SUCCESS 
 8050 tests run, 5 skipped, 0 failed.
--none--

**Comment by asfgit** [comment]
FAILURE
 7942 tests run, 5 skipped, 1 failed.
--none--

**Comment by rhauch** [comment]
retest this please

**Comment by asfgit** [comment]
FAILURE
 No test results found.
--none--

**Comment by asfgit** [comment]
SUCCESS 
 8050 tests run, 5 skipped, 0 failed.
--none--

**Comment by asfgit** [comment]
SUCCESS 
 8050 tests run, 5 skipped, 0 failed.
--none--

**Comment by tedyu** [comment]
@rhauch 
Do you have suggestion on what else should be done ?

**Review by rhauch — APPROVED** [review]
This looks great, @tedyu. Thanks for the contribution.

For other reviewers: the hash code is computed in exactly the same way as before, except it is now computed in the constructor rather than every time `hashCode()` is called. Also, the _order_ that objects are compared in `equals()` has changed, but this order doesn't affect the outcome as all the same fields must be equal in order for two Schemas to be equals.

**Comment by rhauch** [comment]
@kkonstantine, care to review before having a committer review and merge?

File: connect/api/src/main/java/org/apache/kafka/connect/data/ConnectSchema.java
Comment (kkonstantine) [inline connect/api/src/main/java/org/apache/kafka/connect/data/ConnectSchema.java:288]: Are these 3 lines an artifact of code reformatting? They don't seem to correspond to changes other than the reordering itself. The previous order matches the variable declaration order, so maybe we want to leave this as-is.

File: connect/api/src/main/java/org/apache/kafka/connect/data/ConnectSchema.java
Comment (kkonstantine) [inline connect/api/src/main/java/org/apache/kafka/connect/data/ConnectSchema.java]: Traditionally we don't overuse acronyms for variables. Should we call it `hash` or `hashCode` to make really obvious what this variable stands for? Also, to be even more _nitpicky_ given the small range of changes, I believe we add a comment on its own line (see above comment line for `doc` variable).

**Review by kkonstantine — COMMENTED** [review]
Looks good, given that there's no mutating operation on the collections of the class. 

2-3 nitpicks that'd be nice to address since the changes are very small.

File: connect/api/src/main/java/org/apache/kafka/connect/data/ConnectSchema.java
Comment (tedyu) [reply connect/api/src/main/java/org/apache/kafka/connect/data/ConnectSchema.java:288]: The reordering is for comparing primitive data type fields first.

Comment (kkonstantine) [reply connect/api/src/main/java/org/apache/kafka/connect/data/ConnectSchema.java:288]: Are you implying cache efficiency? 

I'd imagine the order to matter if one of the fields had more chances to return `false` than the others. But I don't think it's worth thinking about this here. Feel free to leave this as you have it. When I saw it, I just thought your editor performed the reordering.

**Review by kkonstantine — APPROVED** [review]
Thanks, good catch.

**Comment by asfgit** [comment]
FAILURE
 No test results found.
--none--

**Comment by asfgit** [comment]
SUCCESS 
 8051 tests run, 5 skipped, 0 failed.
--none--

**Comment by asfgit** [comment]
SUCCESS 
 8050 tests run, 5 skipped, 0 failed.
--none--

**Comment by tedyu** [comment]
Can a committer review this ?

Thanks

**Comment by rhauch** [comment]
@ewencp would you care to review and merge into `trunk`?

**Comment by tedyu** [comment]
Ping @ewencp

**Comment by tedyu** [comment]
Ping @ewencp

File: connect/api/src/main/java/org/apache/kafka/connect/data/ConnectSchema.java
Comment (ewencp) [inline connect/api/src/main/java/org/apache/kafka/connect/data/ConnectSchema.java]: Do we actually want this to always be precomputed or just on demand (i.e. store as an `Integer` that's `null` until the first time `hashCode` is called)?

File: connect/api/src/main/java/org/apache/kafka/connect/data/ConnectSchema.java
Comment (ewencp) [reply connect/api/src/main/java/org/apache/kafka/connect/data/ConnectSchema.java:288]: @kkonstantine I think the point is to avoid walking nested data structures that would require pointer chasing, so essentially yes it is a cache efficiency thing. If we really wanted to optimize this we would want to make sure everything that can be inlined in the class is checked first, i.e. `optional`, `type`, and `version`, before other simple fields like `name` and `doc`.

File: connect/api/src/main/java/org/apache/kafka/connect/data/ConnectSchema.java
Comment (rhauch) [reply connect/api/src/main/java/org/apache/kafka/connect/data/ConnectSchema.java]: Avro's `Schema` computes and caches the hash code [on demand](https://github.com/apache/avro/blob/master/lang/java/avro/src/main/java/org/apache/avro/Schema.java#L354-L358). Doing that does add overhead (non-synchronized computing and setting of the hash code to a non-volatile field might mean it's computed a few times), so considering that Connect will often need the hash code is it really worth optimizing for that case? Certainly the Avro converter will use the `hashCode()` method for every `ConnectSchema` it sees.

I guess I'm in favor of precomputing because it is simpler and will be needed much of the time anyway.

Comment (tedyu) [reply connect/api/src/main/java/org/apache/kafka/connect/data/ConnectSchema.java]: +1 on precomputing hash code.

**Comment by tedyu** [comment]
@ewencp 
Do you have other comment(s) ?

File: connect/api/src/main/java/org/apache/kafka/connect/data/ConnectSchema.java
Comment (ewencp) [reply connect/api/src/main/java/org/apache/kafka/connect/data/ConnectSchema.java]: I asked because Connect doesn't make any guarantees about thread safety of that class in the first place (within the framework, everything is sufficiently locked, though people may be implicitly making assumptions in connectors since it was intended to be immutable anyway). We're making micro-optimizations here without benchmarks, so we really do need to take care.

I can think of cases we're definitely increasing overhead here (e.g. any transformation that filters out data would probably mean schema hash is never computed whereas now it is). I can't think of any cases in the connectors we've built where you would end up double computing (they pretty much all do any sort of schema hashing in a single thread because the whole point of tasks in connect is to handle parallelism at the fmwk level automatically for you, and any threading usually is internal to the library handling the connection to the other data storage system; the caching in converters would also generally be safe except for static schemas, and in that case the overhead is not worth worrying about and I think truly static schemas in connectors are pretty rare anyway).

Comment (tedyu) [reply connect/api/src/main/java/org/apache/kafka/connect/data/ConnectSchema.java]: If I use Integer (initially null) to represent the hash code, each subsequent hashCode() would involve a check (against null) and unboxing.
Just want to confirm that it is Okay going this route.

**Comment by asfgit** [comment]
FAILURE
 7973 tests run, 5 skipped, 1 failed.
--none--

**Comment by asfgit** [comment]
FAILURE
 7973 tests run, 5 skipped, 1 failed.
--none--

**Comment by asfgit** [comment]
SUCCESS 
 8081 tests run, 5 skipped, 0 failed.
--none--

**Comment by tedyu** [comment]
@ewencp
Mind taking another look ?

**Comment by ewencp** [comment]
LGTM

**Pull request closed without merging**
